google/api/field_behavior.proto

```diff
@@ -0,0 +1,59 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package google.api;
+
+option go_package = "google.golang.org/genproto/googleapis/api/annotations;annotations";
+option java_multiple_files = true;
+option java_outer_classname = "FieldBehaviorProto";
+option java_package = "com.google.api";
+option objc_class_prefix = "GAPI";
+
+
+// An indicator of the behavior of a given field (for example, that a field
+// is required in requests, or given as output but ignored as input).
+// This DOES NOT change the behavior in protocol buffers itself; it only
+// denotes the behavior and may affect how API tooling handles the field.
+enum FieldBehavior {
+  FIELD_BEHAVIOR_UNSPECIFIED = 0;
+
+  // Specifically denotes a field as optional.
+  // Because fields are optional by default, this annotation is unnecessary;
+  // however, it may be provided if doing so is useful for humans reading
+  // the file.
+  OPTIONAL = 1;
+
+  // Denotes a field as required.
+  // This indicates that the field *must* be provided as part of the request,
+  // and failure to do so will cause an error (usually `INVALID_ARGUMENT`).
+  REQUIRED = 2;
+
+  // Denotes a field as output only.
+  // This indicates that the field is provided in responses, but including the
+  // field in a request does nothing (the server *must* ignore it and
+  // *must not* throw an error as a result of the field's presence).
+  OUTPUT_ONLY = 3;
+
+  // Denotes a field as input only.
+  // This indicates that the field is provided in requests, and the
+  // corresponding field is not included in output.
+  INPUT_ONLY = 4;
+
+  // Denotes a field as immutable.
+  // This indicates that the field may be set once in a request to create a
+  // resource, but may not be changed thereafter.
+  IMMUTABLE = 5;
```

Comment:
I think I recall hearing that only the `REQUIRED` enum has to be implemented by the gapic-generators. Who's going to implement handling the `IMMUTABLE` field?

Reply:
At the moment, nobody. It is likely that other tooling may draw from this in the future (Docgen is a good possibility.)

Reply:
Is there any tools that are slated to implement anything on top of the extra enums? I'd think that adding them when their needed would be better.

Reply:
Would it be OK to wait until someone needs the other enums to define the other enums? Otherwise it might mislead API owners to use these new enums because they think they're supported but in actuality no one is supported the new enums.

Reply:
> Would it be OK to wait until someone needs the other enums to define the other enums? 

In general I agree with this principle, but am confident about this case. We have had these five well-understood field behaviors for some time, and we already have a documentation convention around them (which this would allow us to replace).

> Otherwise it might mislead API owners to use these new enums because they think they're supported but in actuality no one is supported the new enums.

That is okay, so long as they are accurately describing their API surface. These annotations are about _what the field is_, not any particular desired behavior.